sql/descriptor_mutation_test.go

```diff
@@ -642,6 +644,334 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR, i CHAR, INDEX foo (i, v), FAMIL
+
+	mTest := mutationTest{
+		T:         t,
+		kvDB:      kvDB,
+		sqlDB:     sqlDB,
+		tableDesc: tableDesc,
+	}
+
+	starQuery := `SELECT * FROM t.test`
+	indexQuery := `SELECT v FROM t.test@foo`
+	// See the effect of the operations depending on the state.
+	for _, state := range []sqlbase.DescriptorMutation_State{sqlbase.DescriptorMutation_DELETE_ONLY, sqlbase.DescriptorMutation_WRITE_ONLY} {
+		// Init table with some entries.
+		if _, err := sqlDB.Exec(`TRUNCATE TABLE t.test`); err != nil {
+			t.Fatal(err)
+		}
+		initRows := [][]string{{"a", "z"}, {"b", "y"}}
+		for _, row := range initRows {
+			if _, err := sqlDB.Exec(`UPSERT INTO t.test VALUES ($1, $2)`, row[0], row[1]); err != nil {
+				t.Fatal(err)
+			}
+		}
+		_ = mTest.checkQueryResponse(starQuery, initRows)
+		// Index foo is visible.
+		_ = mTest.checkQueryResponse(indexQuery, [][]string{{"y"}, {"z"}})
+
+		// Index foo is invisible once it's a mutation.
+		mTest.writeIndexMutation("foo", sqlbase.DescriptorMutation{State: state})
+		if _, err := sqlDB.Query(indexQuery); !testutils.IsError(err, `index "foo" not found`) {
+			t.Fatal(err)
+		}
+
+		// Insert a new entry.
+		if _, err := sqlDB.Exec(`UPSERT INTO t.test VALUES ('c', 'x')`); err != nil {
+			t.Fatal(err)
+		}
+		_ = mTest.checkQueryResponse(starQuery, [][]string{{"a", "z"}, {"b", "y"}, {"c", "x"}})
+
+		// Make index "foo" live so that we can read it.
+		mTest.makeMutationsActive()
+		if state == sqlbase.DescriptorMutation_DELETE_ONLY {
+			// "x" didn't get added to the index.
+			_ = mTest.checkQueryResponse(indexQuery, [][]string{{"y"}, {"z"}})
+		} else {
+			// "x" got added to the index.
+			_ = mTest.checkQueryResponse(indexQuery, [][]string{{"x"}, {"y"}, {"z"}})
+		}
+
+		// Make "foo" a mutation.
+		mTest.writeIndexMutation("foo", sqlbase.DescriptorMutation{State: state})
+		// Update.
+		if _, err := sqlDB.Exec(`UPSERT INTO t.test VALUES ('c', 'w')`); err != nil {
+			t.Fatal(err)
+		}
+		// Update "v" to its current value "z" in row "a".
+		if _, err := sqlDB.Exec(`UPSERT INTO t.test VALUES ('a', 'z')`); err != nil {
+			t.Fatal(err)
+		}
+		_ = mTest.checkQueryResponse(starQuery, [][]string{{"a", "z"}, {"b", "y"}, {"c", "w"}})
+
+		// Make index "foo" live so that we can read it.
+		mTest.makeMutationsActive()
+		if state == sqlbase.DescriptorMutation_DELETE_ONLY {
+			// updating "x" -> "w" is a noop on the index,
+			// updating "z" -> "z" results in "z" being deleted from the index.
+			_ = mTest.checkQueryResponse(indexQuery, [][]string{{"y"}, {"z"}})
+		} else {
+			// updating "x" -> "w" results in the index updating from "x" -> "w",
+			// updating "z" -> "z" is a noop on the index.
+			_ = mTest.checkQueryResponse(indexQuery, [][]string{{"w"}, {"y"}, {"z"}})
+		}
+	}
+}
+
+// TestUpsertWithUniqueColumnMutation tests UPSERT operation while an index
+// mutation refers to a column mutation.
+func TestUpsertWithUniqueColumnMutation(t *testing.T) {
```

Comment:
I'm finding it incredibly difficult to spot the differences between these 6 tests. Seems worth investigating another abstraction, something that (as much as possible) only leaves the bits that are different in the various test functions